File: liam/scope.py

#This is global scope variable meaning that anywhere in the program it can be accessed
x = 1

def foo4():

    fruit = input("Enter a fruit ")

    if fruit == 'banana':
        x = 1337
        print("foo4's x variable = ", x )
    else:
        #Program is going to crash here, I want you to tell me why - might require a little research
        print("foo4's x variable = ", x )
        print("next time type in banana")

def foo3(x):

    # this x is going to be local to foo3, only code written in foo3 function can access this x
    x += 1
    print("Foo3's x = ", x)

def foo2():

    #As you can notice foo2 doesn't have an x variable, so it will then look upwards to the global scope to find an x variable to print
    print("foo 2's x = ", x)

def foo():
    
    # this x is going to be local to foo, only code written in foo function can access this x
    x = 2
    print("foo's x = ", x)
    foo2()
    foo3(x)

def main():
    
    # this x is going to be local to main, only code written in the maing function can access this x
    x = 3
    print("Gloabl x = ", x)

    foo()
    foo4()


main()

File: liam/scope.cpp
#include<iostream>
#include<string>

using namespace std;

int x = 3;

void foo4(){

    string fruit;
    cout << "Enter a fruit\t";
    cin >> fruit;

    if(fruit != "banana"){
        cout << "The value of x in foo4  is= " << x << " with a memory address of " << &x << endl;
        cout << "Next time eneter banana" << endl;
    }
    else{
        cout << "The value of x in foo4  is= " << x << " with a memory address of " << &x << endl;
    }
}

void foo3(int x){

    x++;

        cout << "The value of x in foo3  is= " << x << " with a memory address of " << &x << endl;
}

void foo2(){

        cout << "The value of x in foo2  is= " << x << " with a memory address of " << &x << endl;
}

void foo(){

    int x = 2;

        cout << "The value of x in foo  is= " << x << " with a memory address of " << &x << endl;

    foo2();
    foo3(x);
}

int main(){

    cout << "The global x = " << x << " with a memory address of " << &x << endl;
    foo();
    foo4();
    
    return 0;
}